theano/tensor/opt.py

```diff
@@ -1431,29 +1431,93 @@ def local_useless_elemwise(node):
     mul(x) -> x
     add(x) -> x
     identity(x) -> x
+    and(x,1) -> x
+    and(x,0) -> zeros_like(x)
+    or(x,0) -> x
+    or(x,1) -> ones_like(x)
+    xor(x,x) -> zeros_like(x)
 
     """
     if isinstance(node.op, T.Elemwise):
+        def zeros_like(node, in_idx):
+            # it is the same var in the graph. That will always be true
+            return [T.fill(node.inputs[in_idx],
+                           T.constant(0.0, dtype=node.outputs[0].type.dtype))]
+
+        def ones_like(node, in_idx):
+            # it is the same var in the graph. That will always be true
+            return [T.fill(node.inputs[in_idx],
+                           T.constant(1.0, dtype=node.outputs[0].type.dtype))]
+
         if node.op.scalar_op == theano.scalar.eq and len(node.inputs) == 2:
             if node.inputs[0] == node.inputs[1]:
-            # it is the same var in the graph. That will always be true
-                return [T.fill(node.inputs[0],
-                               T.constant(1.0,
-                                          dtype=node.outputs[0].type.dtype))]
+                return ones_like(node, 0)
         elif node.op.scalar_op == theano.scalar.neq and len(node.inputs) == 2:
             if node.inputs[0] == node.inputs[1]:
-            # it is the same var in the graph. That will always be false
-                return [T.fill(node.inputs[0],
-                               T.constant(0.0,
-                                          dtype=node.outputs[0].type.dtype))]
+                return zeros_like(node, 0)
         elif node.op.scalar_op == theano.scalar.mul and len(node.inputs) == 1:
             return [node.inputs[0]]
         elif node.op.scalar_op == theano.scalar.add and len(node.inputs) == 1:
             return [node.inputs[0]]
-
-        elif (node.op.scalar_op == theano.scalar.identity
-            and len(node.inputs) == 1):
+        elif (
+            node.op.scalar_op == theano.scalar.identity
+            and len(node.inputs) == 1
+        ):
             return [node.inputs[0]]
+        elif (
+            isinstance(node.op.scalar_op, scalar.AND)
+            and len(node.inputs) == 2
+        ):
+            if (
+                isinstance(node.inputs[0], T.TensorConstant)
+                and T.extract_constant(node.inputs[0]) == 1
+            ):
+                return [node.inputs[1]]
+            if (
+                isinstance(node.inputs[1], T.TensorConstant)
+                and T.extract_constant(node.inputs[1]) == 1
+            ):
+                return [node.inputs[0]]
+            if (
+                isinstance(node.inputs[0], T.TensorConstant)
+                and T.extract_constant(node.inputs[0]) == 0
+            ):
+                return zeros_like(node, 1)
+            if (
+                isinstance(node.inputs[1], T.TensorConstant)
+                and T.extract_constant(node.inputs[1]) == 0
+            ):
```

Comment:
You call extract_constant 2 times with the same input var, can you refactor that to prevent that?